### ISSUE TITLE
Implementing `current()`

### ISSUE THREAD

**Issue by Hipska** [antchfx/xpath#73]
I'm wondering why the `current()` function hasn't been implemented, and if it would make much time to do so.

I'm currently in the situation where I would need a query like `//cd[@title=current()/@ref]`. The node navigator used for the expression evaluate isn't the document root, but a previously selected node.

**Comment by Hipska** [comment]
@zhengchun?

**Comment by zhengchun** [comment]
`current()` is only supported by XSLT. You can try this `//cd[@title=@ref]`

**Comment by Hipska** [comment]
Doesn't that only select `cd` nodes of which title attribute equals their ref attribute?

**Comment by Hipska** [comment]
Scrap that, it seems it didn't work, it always picks the first one in this case: `../../measType[@p=@p]`

**Comment by zhengchun** [comment]
Hello, Need a sample document to test.

or try this `../../measType[@p]`

**Comment by Hipska** [comment]
That gives the same result. Its strange, as I was really under the impression your solution first worked..

Anyway, this is a simplified version of the file:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet type="text/xsl" href="MeasDataCollection.xsl"?>
<measCollecFile xmlns="http://www.3gpp.org/ftp/specs/archive/32_series/32.435#measCollec">
	<measData>
		<managedElement localDn="foobar.example.com" />
		<measInfo measInfoId="metric_name">
			<measType p="1">field1</measType>
			<measType p="2">field2</measType>
			<measValue measObjLdn="identifier">
				<r p="1">31854</r>
				<r p="2">159773</r>
			</measValue>
		</measInfo>
	</measData>
</measCollecFile>
```

And the selected node is each time `r`, so I need an xpath to find out the field name of it.

**Comment by zhengchun** [comment]
like this `//r[@p]/ancestor::measInfo/measType[@p]`

**Comment by Hipska** [comment]
That's even worse, now all metrics have field name `field1`.

The Path is run on each `r` node. So I think only `../../measType[@p=current()/@p]` will work..

**Comment by zhengchun** [comment]
using the nested query to solve. like this

```go
for _,n:=range xmlquery.Find(doc,"//r[@p]")
{
    for _,nn:=range xmlquery.Find(n,`../../measType[@p='`+n.SelectAttr("p")+`']`){}
}
```

**Comment by Hipska** [comment]
Well, yeah. I can't change the code. I'm using telegraf: https://github.com/influxdata/telegraf/blob/6dc2b99549b3fc4684295ebcd2975cee96a1fa28/plugins/parsers/xpath/parser.go#L419-L447

**Comment by zhengchun** [comment]
I tested `//r[@p=../../measType/@p]` and it only return the first one node, this is a bug.

**Comment by zhengchun** [comment]
`//r[@p=../../measType/@p]`  should be your expression to sove your problem.

**Comment by Hipska** [comment]
No, I have an `r` and need an xpath to find it’s measType. 

Your xpath selects every r that has a corresponding measType.

**Comment by Hipska** [comment]
@zhengchun any update on this? Can I help to fix this?

**Comment by zhengchun** [comment]
@Hipska , You can try to fix this and submit PR.

**Comment by zhengchun** [comment]
Sorry,I no idea how to implement `Current()` now, It's a XSLT function.

**Comment by Hipska** [comment]
Oh, but it seems you fixed it?

**Comment by zhengchun** [comment]
I just fixed this expression `a[@b=../@b]`, fixed bug with parent query on current node in the filter query.

**Comment by Hipska** [comment]
Hmm, then this playground should return 4 nodes, right? https://go.dev/play/p/qeMkl0NqoOT

**Comment by zhengchun** [comment]
@Hipska , The actually result from your go code execute is correct, your expect result is not right.

```
	for i, n := range xmlquery.Find(doc, "//r[@p=../../measType/@p]") {
		fmt.Printf("#%d %s\n", i, n.OutputXML(true))
		for _, nn := range xmlquery.Find(n, `../../measType[@p=./@p]`) {  -- Here
			fmt.Printf("\t%s\n", nn.OutputXML(true))
		}
	}
```

**xmlquery.Find(n, `../../measType[@p=./@p]`)**, HERE, you tell xpath to start search from the specified node `n`, xpath will according on your `../../measType[@p=./@p]` to matching. and `@p=./@p` just telling xpath target node have `@p` attribute, not your expect `@p=n[@p]'.

You should change your query from `xmlquery.Find(n, "../../measType[@p=./@p]") ` to `xmlquery.Find(n, "../../measType[@p=“+ n.Attr("p") +”]“) `

You can try take your xpath expression execute on the another program language and see the result.

**Comment by zhengchun** [comment]
So, what is your expected xpath if we can implement the `current()` function.

 In the above example, if you can nested `for...loop` why you can't change xpath expression to replace `../../measType[@p=./@p]` with `../../measType[@p=“+ n.Attr("p") +”]`?

You can provide an example and what is you expected result.

**Comment by Hipska** [comment]
Hi, sorry to be so late on this..

In the telegraf config, I'm only able to pass a string, see earlier shared code. So I need a single string/xpath that returns the expected `measType` node. `../../measType[@p=“+ n.Attr("p") +”]` is not possible and will not work.

**Comment by zhengchun** [comment]
I find a solution that saving the query node 'n' as a local variable used
by 'current()' function in the query context, like this
***@***.***=current ***@***.***")'。

Thomas Casteleyn ***@***.***> 于 2023年2月14日周二 下午4:44写道：

**Comment by zhengchun** [comment]
I find a solution that saving the query node 'n' as a local variable used 
by `current()` function in the query context, like this `jsonquery.Find(n,"../../measType[@p=current()/@p]")`

**Comment by Hipska** [comment]
Yeah, the whole reason of this issue is because of the `current()` function is not implemented:
> panic: not yet support this function current()

**Comment by Hipska** [comment]
@zhengchun can you point me to where I could implement that?

**Comment by zhengchun** [comment]
Implementing `current()` is a big work, you need to save the current matched node and then used by `current()`  function. One possible approach is to use [context package](https://pkg.go.dev/context) to save the current node.

**Comment by Hipska** [comment]
Okay, but *where* could I start implementing it?

**Comment by zhengchun** [comment]
1. Add a new `current()` function: https://github.com/antchfx/xpath/blob/14e235f5e127e9a750a9b6b2f76a27220d1cfd18/build.go#L335
2. Save the current node in `query` interface, like `child::*`, `descendant::*`, `parent::*`...:
https://github.com/antchfx/xpath/blob/14e235f5e127e9a750a9b6b2f76a27220d1cfd18/query.go#L53

These are only my guesses and cannot be guaranteed to be right.
